File: src/static/rules/SOL-001-unchecked-call.ts
import { SupportedLanguage } from "../../engine/types.js";
import type { Rule, FindingInstance, RuleContext } from "../../engine/types.js";
import type { Node } from "web-tree-sitter";

/**
 * SOL-001: Unchecked Low-Level Call
 *
 * Detects .call(), .delegatecall(), .send() whose return value is discarded.
 * The parent node is an expression_statement (no assignment), meaning the
 * bool success is never checked — the call can silently fail.
 *
 * SWC-104
 */
function createRule(): Rule {
    let findings: FindingInstance[] = [];

    return {
        id: 'SOL-001',
        severity: 'high',
        title: 'Unchecked low-level call return value',
        description: 'Low-level calls (.call, .delegatecall, .send) return a boolean success flag. Discarding it means silent failures — the caller assumes success when the call may have reverted.',
        kind: 'smell',
        appliesTo: {
            languages: [SupportedLanguage.Solidity],
            domains: ['on-chain'],
        },

        enter(node: Node, ctx: RuleContext) {
            if (!ctx.trait.isExternalCall(node)) return;

            // If the call is inside an expression_statement (possibly wrapped in
            // an 'expression' node), the return value is discarded.
            let ancestor = node.parent;
            while (ancestor && ancestor.type === 'expression') {
                ancestor = ancestor.parent;
            }
            if (!ancestor || ancestor.type !== 'expression_statement') return;

            findings.push({
                location: {
                    file: ctx.currentFile,
                    line: node.startPosition.row + 1,
                    col: node.startPosition.column,
                },
                snippet: node.text.length > 120 ? node.text.slice(0, 117) + '...' : node.text,
            });
        },

        finalize() { return findings; },
        reset() { findings = []; },
    };
}

export default createRule();

File: src/rules/SOL-001-unchecked-call.lua
rule = {
    id = "SOL-001",
    name = "unchecked-call",
    severity = "high",
    type = "scope",
    confidence = "smell",
    languages = {"solidity"},
    description = "Low-level .call(), .send(), and .delegatecall() return a bool success flag. Discarding it silently swallows reverts.",
}

local LOW_LEVEL = { call = true, send = true, delegatecall = true }

-- Solidity grammar wraps sub-expressions in `expression` nodes.
local function unwrap(h)
    if h and ast.type(h) == "expression" then
        local kids = ast.named_children(h)
        return kids[1]
    end
    return h
end

function enter(node, ctx)
    if node.kind ~= "call_expression" then return end

    local h = node.handle
    local fn_inner = unwrap(ast.child_by_field(h, "function"))
    if fn_inner == nil then return end
    if ast.type(fn_inner) ~= "member_expression" then return end

    local prop = ast.child_by_field(fn_inner, "property")
    if prop == nil then return end
    if not LOW_LEVEL[ast.text(prop)] then return end

    -- The return value is discarded when the call_expression sits inside
    -- expression > expression_statement (Solidity wraps in expression node).
    local parent = ast.parent(h)
    local grandparent = parent and ast.parent(parent) or nil
    if grandparent == nil then return end
    if ast.type(grandparent) ~= "expression_statement" then return end

    report.hit({
        file = ctx.current_file,
        line = node.line,
        node_text = ast.text(prop),
    })
end

function exit(node, ctx) end
